llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -4698,6 +4698,20 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *bv, Function *f, Bas
     auto trueSuccessor = cast<BasicBlock>(transValue(br->getTrueLabel(), f, bb));
     auto falseSuccessor = cast<BasicBlock>(transValue(br->getFalseLabel(), f, bb));
     auto bc = BranchInst::Create(trueSuccessor, falseSuccessor, c, bb);
+
+    // OpBranchConditional can branch with OpUndef as condition. The selected jump target is undefined.
+    // In LLVM IR, BranchInst with undef value is fully UB.
+    // LLVM will mark the successors as unreachable, so later SimplifyCFG passes will cause
+    // unwanted removal of branches. By using a freeze instruction as condition, we make sure that LLVM
+    // will operate on a fixed value.
+    if (isa<UndefValue>(c)) {
+      auto builder = getBuilder();
```

Comment:
Nit: I know Nicolai suggested using IRBuilder, but I think (hope) he meant converting all of the "case OpBranchConditional" to use it. Using it just for inserting this one instruction makes the code more complicated for no benefit.

Reply:
Doesn't it make sense to use the builder to get the earliest insertion point after the allocas?

Reply:
It is simpler to insert the freeze immediately before the branch.